File: semestre-1/disciplina-3-arquitetura-de-dados/unidade-2/aula-3/aula-3.py
# aula-3.py
# Exemplo de Armazenamento de Dados: SQL em Ação com Python
# Demonstra as categorias DDL (CREATE), DML (INSERT, SELECT) e DTL (COMMIT)
# e serve como base para a conexão a um serviço como o SQL Azure na Nuvem.

import sqlite3
import os

# --------------------------------------------------------------------------------------
# 1. Conexão ao "Castelo" (Banco de Dados SQL)
# --------------------------------------------------------------------------------------
# Usamos o SQLite, um banco de dados local embutido, para simular um SGBD Relacional
# A conexão com um SQL Azure ou PostgreSQL seria muito similar, apenas mudando a string de conexão.

DB_NAME = 'tesouro_dados_aula3.db'

# Remove o arquivo do banco de dados se ele já existir para garantir um teste limpo
if os.path.exists(DB_NAME):
    os.remove(DB_NAME)

try:
    # Cria uma conexão e um cursor (o "operário" que envia comandos)
    conn = sqlite3.connect(DB_NAME)
    cursor = conn.cursor()
    print(f"✅ Conexão estabelecida com sucesso ao banco de dados: {DB_NAME}")

    # ----------------------------------------------------------------------------------
    # 2. DDL (Data Definition Language) - O Arquiteto constrói a tabela
    # ----------------------------------------------------------------------------------
    # Criamos a tabela 'Projetos_Nuvem' com colunas definidas (estrutura)
    print("\n-- Executando DDL: Criando a tabela 'Projetos_Nuvem' (CREATE TABLE)")
    cursor.execute("""
    CREATE TABLE Projetos_Nuvem (
        id INTEGER PRIMARY KEY,
        nome_projeto TEXT NOT NULL,
        plataforma_nuvem TEXT NOT NULL,
        modelo_servico TEXT, -- IaaS, PaaS, SaaS
        custo_mensal REAL
    );
    """)
    print("📋 Tabela 'Projetos_Nuvem' criada. Estrutura definida!")

    # ----------------------------------------------------------------------------------
    # 3. DML (Data Manipulation Language) - Os Pedreiros inserem os dados
    # ----------------------------------------------------------------------------------
    projetos = [
        ('Data Lake', 'Azure', 'PaaS', 150.00),
        ('Website Institucional', 'AWS', 'IaaS', 45.50),
        ('Email Corporativo', 'GCP', 'SaaS', 25.00),
        ('Análise de Logs', 'Azure', 'PaaS', 89.90)
    ]

    print("\n-- Executando DML: Inserindo novos registros (INSERT)")
    cursor.executemany("INSERT INTO Projetos_Nuvem (nome_projeto, plataforma_nuvem, modelo_servico, custo_mensal) VALUES (?, ?, ?, ?)", projetos)
    print(f"➕ {cursor.rowcount} registros inseridos com sucesso.")

    # ----------------------------------------------------------------------------------
    # 4. DTL (Data Transaction Language) - O Notário garante a persistência
    # ----------------------------------------------------------------------------------
    # O COMMIT garante que as inserções sejam salvas permanentemente no disco.
    print("-- Executando DTL: Confirmando a transação (COMMIT)")
    conn.commit()
    print("✅ Transação confirmada. Dados salvos.")

    # ----------------------------------------------------------------------------------
    # 5. DML (Data Manipulation Language) - O Pedreiro consulta os dados
    # ----------------------------------------------------------------------------------
    print("\n-- Executando DML: Consultando dados do Azure (SELECT com Predicado)")
    cursor.execute("SELECT nome_projeto, modelo_servico FROM Projetos_Nuvem WHERE plataforma_nuvem = 'Azure'")

    print("\nProjetos na Azure:")
    for row in cursor.fetchall():
        print(f"  - Projeto: {row[0]}, Modelo: {row[1]}")

    # Consultando Funções Agregadas (Resumindo os dados)
    print("\n-- Consultando Funções Agregadas (AVG, SUM, COUNT)")
    cursor.execute("SELECT AVG(custo_mensal), SUM(custo_mensal), COUNT(id) FROM Projetos_Nuvem")
    media, soma, total = cursor.fetchone()

    print(f"📊 Análise de Custos:")
    print(f"  - Custo Mensal Médio dos Projetos: R$ {media:.2f}")
    print(f"  - Custo Mensal Total (SUM): R$ {soma:.2f}")
    print(f"  - Número Total de Projetos (COUNT): {total}")


except sqlite3.Error as e:
    print(f"❌ Ocorreu um erro no SQL: {e}")
    # O ROLLBACK seria chamado aqui em um cenário real de erro (DTL)
    if 'conn' in locals() and conn:
        conn.rollback()
        print("↩️ Transação desfeita (ROLLBACK) devido ao erro.")

finally:
    # 6. Finalização - Fechando a porta do castelo
    if 'conn' in locals() and conn:
        conn.close()
        print("\n✅ Conexão SQL encerrada.")

# NOTA SOBRE NUVEM:
# Para conectar ao SQL Azure, seria necessário instalar o driver 'pyodbc'
# e mudar a linha de conexão para algo como:
# conn = pyodbc.connect('DRIVER={ODBC Driver 17 for SQL Server};SERVER=...;DATABASE=...;UID=...;PWD=...')
# O resto do código (DDL, DML, DTL) permaneceria o mesmo!

File: semestre-1/disciplina-3-arquitetura-de-dados/unidade-2/aula-3/Aula3.java
// Exemplo de Armazenamento de Dados em Java: Conexão JDBC e Comandos SQL
// Demonstra a interação robusta de um sistema corporativo (Java) com o SQL,
// um cenário comum ao se conectar a serviços como SQL Server ou SQL Azure.

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;

public class Aula3 {

    // URL de Conexão: Simulação de um banco de dados SQLite (como o SQL Azure ou MySQL gerenciado)
    // Para Azure, a URL seria: "jdbc:sqlserver://<serverName>.database.windows.net:1433;databaseName=..."
    private static final String JDBC_URL = "jdbc:sqlite:auladb.db";
    
    // NOTA: Para rodar este código, você precisará do driver JDBC do SQLite (sqlite-jdbc).

    public static void main(String[] args) {
        Connection conn = null;

        try {
            // 1. Conexão (Iniciando a comunicação com o Castelo SQL)
            // Simula o login na plataforma Azure/serviço SQL.
            conn = DriverManager.getConnection(JDBC_URL);
            System.out.println("✅ Conexão JDBC estabelecida com sucesso.");
            conn.setAutoCommit(false); // Inicia o conceito de Transação (DTL)

            // 2. DDL (Data Definition Language) - Criação da Estrutura
            criarTabela(conn);

            // 3. DML (Data Manipulation Language) - Inserção de Dados
            inserirProjetos(conn);

            // 4. DTL (Data Transaction Language) - Confirmação
            conn.commit();
            System.out.println("✅ Transação confirmada (COMMIT). Dados salvos permanentemente.");

            // 5. DML (Data Manipulation Language) - Consulta e Análise
            consultarProjetosAzure(conn);
            
        } catch (SQLException e) {
            System.err.println("❌ Erro SQL detectado: " + e.getMessage());
            // 6. DTL (Reversão em caso de erro)
            if (conn != null) {
                try {
                    conn.rollback();
                    System.out.println("↩️ Transação desfeita (ROLLBACK) devido ao erro.");
                } catch (SQLException ex) {
                    System.err.println("Erro ao reverter: " + ex.getMessage());
                }
            }
        } finally {
            // 7. Finalização (Fechando a conexão com a Nuvem)
            if (conn != null) {
                try {
                    conn.close();
                    System.out.println("\n✅ Conexão SQL encerrada.");
                } catch (SQLException e) {
                    System.err.println("Erro ao fechar a conexão: " + e.getMessage());
                }
            }
        }
    }

    // Método que executa o DDL (CREATE TABLE)
    private static void criarTabela(Connection conn) throws SQLException {
        String sqlDDL = """
            CREATE TABLE IF NOT EXISTS Projetos_Nuvem (
                id INTEGER PRIMARY KEY,
                nome_projeto TEXT NOT NULL,
                plataforma_nuvem TEXT NOT NULL,
                modelo_servico TEXT,
                custo_mensal REAL
            );
        """;
        try (Statement stmt = conn.createStatement()) {
            stmt.execute(sqlDDL);
            System.out.println("\n📋 DDL executado: Tabela 'Projetos_Nuvem' criada ou verificada.");
        }
    }

    // Método que executa o DML (INSERT)
    private static void inserirProjetos(Connection conn) throws SQLException {
        String sqlDML = "INSERT INTO Projetos_Nuvem (nome_projeto, plataforma_nuvem, modelo_servico, custo_mensal) VALUES (?, ?, ?, ?)";
        
        try (PreparedStatement pstmt = conn.prepareStatement(sqlDML)) {
            
            // Inserção 1 (Azure PaaS)
            pstmt.setString(1, "Data Warehouse Central");
            pstmt.setString(2, "Azure");
            pstmt.setString(3, "PaaS");
            pstmt.setDouble(4, 300.50);
            pstmt.executeUpdate();

            // Inserção 2 (GCP IaaS)
            pstmt.setString(1, "Máquinas Virtuais de Teste");
            pstmt.setString(2, "GCP");
            pstmt.setString(3, "IaaS");
            pstmt.setDouble(4, 75.20);
            pstmt.executeUpdate();

            System.out.println("\n➕ DML executado: 2 novos registros inseridos.");
        }
    }

    // Método que executa o DML (SELECT) e funções agregadas
    private static void consultarProjetosAzure(Connection conn) throws SQLException {
        String sqlSelect = "SELECT nome_projeto, modelo_servico, custo_mensal FROM Projetos_Nuvem WHERE plataforma_nuvem = 'Azure'";
        
        try (Statement stmt = conn.createStatement();
             ResultSet rs = stmt.executeQuery(sqlSelect)) {
            
            System.out.println("\n-- Executando DML: Consultando Projetos na Azure (SELECT com WHERE)");
            
            double custoTotalAzure = 0;
            int contador = 0;

            while (rs.next()) {
                String nome = rs.getString("nome_projeto");
                String modelo = rs.getString("modelo_servico");
                double custo = rs.getDouble("custo_mensal");
                
                System.out.println("  - Projeto: " + nome + ", Modelo: " + modelo + ", Custo: R$ " + custo);
                
                custoTotalAzure += custo; // Simula Função Agregada (SUM)
                contador++; // Simula Função Agregada (COUNT)
            }
            
            // Cálculo da média
            double custoMedio = contador > 0 ? custoTotalAzure / contador : 0; // Simula Função Agregada (AVG)

            System.out.println("\n📊 Análise (Funções Agregadas):");
            System.out.println("  - Total de Projetos Azure (COUNT): " + contador);
            System.out.println("  - Custo Total Azure (SUM): R$ " + String.format("%.2f", custoTotalAzure));
            System.out.println("  - Custo Médio Azure (AVG): R$ " + String.format("%.2f", custoMedio));
        }
    }
}